File: backend/app/api/v1/validate.py
"""
資料驗證 API 端點
"""
from typing import List
from fastapi import APIRouter, HTTPException, Body
from app.services.automation.data_validator import DataValidator
from app.models.topic import Category
import logging

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/validate", tags=["validate"])

# Service 實例
data_validator = DataValidator()


@router.post("/sources")
async def validate_sources(
    topic_id: str = Body(..., description="主題 ID"),
    sources: List[dict] = Body(..., description="來源列表")
):
    """
    驗證並抓取來源資料
    """
    try:
        validated_sources = []
        failed_sources = []
        
        for source in sources:
            validation_result = await data_validator.validate_and_fetch_source(
                source_url=source.get("url", ""),
                source_name=source.get("name", "")
            )
            
            if validation_result.get("valid"):
                validated_sources.append(validation_result)
            else:
                failed_sources.append({
                    "source": source,
                    "reason": validation_result.get("reason", "驗證失敗")
                })
        
        return {
            "topic_id": topic_id,
            "validated_sources": validated_sources,
            "validation_summary": {
                "total_sources": len(sources),
                "verified_sources": len(validated_sources),
                "failed_sources": len(failed_sources)
            },
            "failed_sources": failed_sources
        }
    except Exception as e:
        logger.error(f"驗證來源失敗: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/topic-consistency")
async def validate_topic_consistency(
    keyword: str = Body(..., description="關鍵字"),
    category: Category = Body(..., description="主題分類"),
    sources: List[dict] = Body(..., description="來源列表")
):
    """
    驗證主題的跨來源一致性
    """
    try:
        validation = await data_validator.validate_topic_consistency(
            keyword=keyword,
            category=category,
            sources=sources
        )
        
        return {
            "valid": validation.get("valid", False),
            "confidence": validation.get("consistency_score", 0.0),
            "consistency_score": validation.get("consistency_score", 0.0),
            "sources_verified": len(sources),
            "is_factual": validation.get("is_factual", False),
            "warnings": [validation.get("warning")] if validation.get("warning") else []
        }
    except Exception as e:
        logger.error(f"驗證一致性失敗: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/source-health/{source_url:path}")
async def check_source_health(source_url: str):
    """
    檢查來源健康度
    """
    try:
        health = await data_validator.check_source_health(source_url)
        return health
    except Exception as e:
        logger.error(f"檢查來源健康度失敗: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/images")
async def validate_images():
    """
    驗證圖片服務配置狀態
    返回各圖片服務 API Key 的可用性
    """
    from app.config import settings
    
    # 使用列表推導式過濾 None，確保只返回實際可用的服務
    available_services = [
        svc for svc in [
            "duckduckgo",  # 總是可用
            "unsplash" if settings.UNSPLASH_ACCESS_KEY else None,
            "pexels" if settings.PEXELS_API_KEY else None,
            "pixabay" if settings.PIXABAY_API_KEY else None,
            "google_custom_search" if (settings.GOOGLE_API_KEY and settings.GOOGLE_SEARCH_ENGINE_ID) else None
        ] if svc
    ]
    
    return {
        "unsplash": bool(settings.UNSPLASH_ACCESS_KEY),
        "pexels": bool(settings.PEXELS_API_KEY),
        "pixabay": bool(settings.PIXABAY_API_KEY),
        "google_api_key": bool(settings.GOOGLE_API_KEY),
        "google_search_engine_id": bool(settings.GOOGLE_SEARCH_ENGINE_ID),
        "duckduckgo": True,  # DuckDuckGo 不需要 API Key
        "available_services": available_services
    }

File: frontend/src/api/validate.ts
/**
 * Validate API
 * 資料驗證相關 API
 */

import { fetchAPI, USE_MOCK, delay } from './client'

/**
 * 驗證來源請求
 */
export interface ValidateSourcesRequest {
  topic_id: string
  sources: Array<{
    url: string
    name: string
  }>
}

/**
 * 驗證來源回應
 */
export interface ValidateSourcesResponse {
  topic_id: string
  validated_sources: Array<{
    valid: boolean
    url: string
    name: string
    content_snippet?: string
    screenshot_url?: string
    fetched_at: string
    reliability: string
  }>
  validation_summary: {
    total_sources: number
    verified_sources: number
    failed_sources: number
  }
  failed_sources: Array<{
    source: {
      url: string
      name: string
    }
    reason: string
  }>
}

/**
 * 驗證一致性請求
 */
export interface ValidateConsistencyRequest {
  keyword: string
  category: 'fashion' | 'food' | 'trend'
  sources: Array<{
    url: string
    name: string
    title?: string
  }>
}

/**
 * 驗證一致性回應
 */
export interface ValidateConsistencyResponse {
  valid: boolean
  confidence: number
  consistency_score: number
  sources_verified: number
  is_factual: boolean
  warnings: string[]
}

/**
 * 來源健康度回應
 */
export interface SourceHealthResponse {
  health_score: number
  status: 'healthy' | 'degraded' | 'unhealthy'
  response_time?: number
  status_code?: number
  error?: string
}

/**
 * Validate API
 */
export const validateAPI = {
  /**
   * 驗證並抓取來源資料
   */
  validateSources: async (data: ValidateSourcesRequest): Promise<ValidateSourcesResponse> => {
    if (USE_MOCK) {
      await delay(1000)
      return {
        topic_id: data.topic_id,
        validated_sources: data.sources.map((source) => ({
          valid: true,
          url: source.url,
          name: source.name,
          fetched_at: new Date().toISOString(),
          reliability: 'high',
        })),
        validation_summary: {
          total_sources: data.sources.length,
          verified_sources: data.sources.length,
          failed_sources: 0,
        },
        failed_sources: [],
      }
    }

    return await fetchAPI<ValidateSourcesResponse>('/validate/sources', {
      method: 'POST',
      body: JSON.stringify(data),
    })
  },

  /**
   * 驗證主題的跨來源一致性
   */
  validateTopicConsistency: async (
    data: ValidateConsistencyRequest
  ): Promise<ValidateConsistencyResponse> => {
    if (USE_MOCK) {
      await delay(500)
      return {
        valid: true,
        confidence: 0.9,
        consistency_score: 0.9,
        sources_verified: data.sources.length,
        is_factual: false,
        warnings: [],
      }
    }

    return await fetchAPI<ValidateConsistencyResponse>('/validate/topic-consistency', {
      method: 'POST',
      body: JSON.stringify(data),
    })
  },

  /**
   * 檢查來源健康度
   */
  checkSourceHealth: async (sourceUrl: string): Promise<SourceHealthResponse> => {
    if (USE_MOCK) {
      await delay(300)
      return {
        health_score: 0.9,
        status: 'healthy',
        response_time: 0.5,
        status_code: 200,
      }
    }

    // URL 編碼
    const encodedUrl = encodeURIComponent(sourceUrl)
    return await fetchAPI<SourceHealthResponse>(`/validate/source-health/${encodedUrl}`)
  },
}

